### PR TITLE
Import DHIS2 Data Elements

### PR DESCRIPTION
**Pull request by kaapstorm**
##### SUMMARY

This is a custom script for importing DHIS2 Data Elements for the "onse-iss" project space. Configuration is hard-coded, and expected to change only once, when the project switches from their testing DHIS2 instance to their production instance.

##### FEATURE FLAG

DHIS2 features are normally behind a feature flag, but this code will only execute for a specific domain, so a feature flag seemed unnecessary.

##### RISK ASSESSMENT / QA PLAN

The branch will be deployed to Staging, and QAed by the project team (not the QA team).

##### PRODUCT DESCRIPTION

N/A

### CODE REVIEW

File: custom/onse/tasks.py [hunk 1]
```diff
@@ -0,0 +1,298 @@
...
+            if data_set_cache[data_set] is None:
+                # No data for this facility. `None` = "We don't know"
+                case_block.update[case_property] = None
+            else:
+                value = 0
+                for data_value in data_set_cache[data_set]:
+                    if data_value['dataElement'] == data_element:
+                        value += int(data_value['value'])
+                case_block.update[case_property] = value
+        yield case_block
```
Comment (esoergel) [inline custom/onse/tasks.py:270]: nit: this function is a little on the complex side.  With a small class, you could separate out the body of the loop and extract the data_set caching business so the case block updates are clearer.  I have no idea what most of these variables refer to, so it'd be nice to see it at a higher level of abstraction.

File: custom/onse/tasks.py [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+    return f"{year}Q{last_quarter}"
+
+
+def save_cases(case_blocks):
+    today = date.today().isoformat()
+    for chunk in chunked(case_blocks, 1000, collection=list):
+        submit_case_blocks(
+            [cb.as_text() for cb in chunk],
+            DOMAIN,
+            xmlns='http://commcarehq.org/dhis2-import',
```
Comment (esoergel) [inline custom/onse/tasks.py:296]: keeping things clean, I see :stuck_out_tongue_closed_eyes:

File: custom/onse/get_dataset_ids.py [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+        auth = (USERNAME, PASSWORD)
+        resp = requests.get(url, auth=auth, verify=False)
+        if resp.status_code == 404:
+            print(f'    # {de_name} ({de_id}) not found')
+            continue
+        data_set_id = resp.json()['dataSetElements'][0]['dataSet']['id']
+        print(f"""    '{case_prop}': DataElement(
+        id='{de_id}',  # {de_name}
+        data_set='{data_set_id}',
+    ),""")
```
Comment (esoergel) [inline custom/onse/get_dataset_ids.py:66]: You could probably also store this stuff as a csv or json file if that's easier.  Might be more straightforward for a non-dev to edit too, if that's a realistic scenario.

File: custom/onse/tasks.py [hunk 4]
```diff
@@ -0,0 +1,298 @@
...
+
+def set_case_updates(
+    connection_settings: ConnectionSettings,
+    case_blocks: Iterable[CaseBlock]
+) -> Iterable[CaseBlock]:
+    """
+    Fetch all the data elements for the data set, because we don't know
+    in advance what category option combos to query for
+    """
+    # e.g. https://play.dhis2.org/dev/api/dataValueSets?orgUnit=jNb63DIHuwU
```
Comment (esoergel) [inline custom/onse/tasks.py:242]: would it be straightforward to use this example as a test case?  It'd be nice to declare what the expected response format is so we can adapt easily if it changes.

**Review by esoergel — APPROVED** [review]
Tasks this infrequent can be kinda scary - what if they fail?  Couple suggestions that might help future-proof this:

* Should this send a status email to someone every time it runs?  That'll prempt questions of "did it run this time?  Were there any failures?  And so on.
* If the task is idempotent wrt case state, you might also consider a management command which calls the task directly.  Then we'd be well equipped to re run it on short notice if necessary.

File: custom/onse/tasks.py (hunk 2 above)
Comment (proteusvacuum) [reply custom/onse/tasks.py:296]: Is this a good idea? I know this doesn't mean anything really, but we do have the openrosa convention currently, and I'm not convinced we don't assume that in places in HQ.

Comment (kaapstorm) [reply custom/onse/tasks.py:220]: Using non-openrosa XMLNS happens quite a lot already, usually to indicate the origin or purpose of the form, so I think it should be OK.

e.g.
```python
COMMTRACK_REPORT_XMLNS = 'http://commcarehq.org/ledger/v1'
DEVICE_LOG_XMLNS = 'http://code.javarosa.org/devicereport'
COMMTRACK_SUPPLY_POINT_XMLNS = 'http://commtrack.org/supply_point'
SMS_XMLNS = 'http://commtrack.org/sms_submission'
XMLNS_OPENMRS = 'http://commcarehq.org/openmrs-integration'
```

File: custom/onse/tasks.py (hunk 4 above)
Comment (kaapstorm) [reply custom/onse/tasks.py]: I'm a little hesitant to write automated tests that hit play.dhis2.org, but I have included [a sample return value in a doctest](https://github.com/dimagi/commcare-hq/pull/28742/commits/4a6e866d77efec93128a0754db878c711840ac92#diff-bcabc4569dc5b1ece3959f73e371f4ebd74b1862361b35347da01b2a88313389R239-R254). Do you think that's enough?

**Comment by kaapstorm** [comment]
Thanks for the great feedback. Updates pushed.

File: custom/onse/tasks.py (hunk 4 above)
Comment (esoergel) [reply custom/onse/tasks.py]: Oh yeah, that's basically what I meant - stubbing in an example response like what would be provided by the API.

**Review by esoergel — APPROVED** [review]
:star:

**Review by esoergel — APPROVED** [review]

**Pull request merged**